ledger/ledger-api-integration-tests/src/test/itsuite/scala/com/digitalasset/platform/tests/integration/ledger/api/ContractKeysIT.scala

```diff
@@ -0,0 +1,465 @@
+              RecordField(value = expected match {
+                case None => Value(Value.Sum.Optional(Optional(None)))
+                case Some(cid) => Value(Value.Sum.Optional(Optional(Some(cid.asContractId))))
+              })
+            )
+        )
+        _ <- ctx.testingHelpers.simpleExercise(
+          testIdsGenerator.testCommandId("SDVl6"),
+          submitter = owner,
+          template = templateIds.showDelegated,
+          contractId = showIdEv.contractId,
+          choice = "ShowIt",
+          arg = Value(Value.Sum.Record(fetchArg)),
+        )
+        _ <- ctx.testingHelpers.simpleExercise(
+          testIdsGenerator.testCommandId("SDVl7"),
+          submitter = delegate,
+          template = templateIds.delegation,
+          contractId = delegationEv.contractId,
+          choice = "FetchDelegated",
+          arg = Value(Value.Sum.Record(fetchArg)),
+        )
+        _ <- if (languageVersion precedes LanguageVersion.checkSubmitterInMaintainers) {
+          ctx.testingHelpers.simpleExercise(
+            testIdsGenerator.testCommandId("SDVl8"),
+            submitter = delegate,
+            template = templateIds.delegation,
+            contractId = delegationEv.contractId,
+            choice = "FetchByKeyDelegated",
+            arg = Value(Value.Sum.Record(lookupArg(Some(delegatedEv.contractId)))),
+          )
+        } else {
+          ctx.testingHelpers.failingExercise(
+            testIdsGenerator.testCommandId("SDVl8"),
+            submitter = delegate,
+            template = templateIds.delegation,
+            contractId = delegationEv.contractId,
+            choice = "FetchByKeyDelegated",
+            arg = Value(Value.Sum.Record(lookupArg(Some(delegatedEv.contractId)))),
+            Code.INVALID_ARGUMENT,
+            "Expected the submitter 'Charlie' to be in maintainers 'Bob'",
+          )
+        }
+        _ <- if (languageVersion precedes LanguageVersion.checkSubmitterInMaintainers) {
+          ctx.testingHelpers.simpleExercise(
+            testIdsGenerator.testCommandId("SDVl9"),
+            submitter = delegate,
+            template = templateIds.delegation,
+            contractId = delegationEv.contractId,
+            choice = "LookupByKeyDelegated",
+            arg = Value(Value.Sum.Record(lookupArg(Some(delegatedEv.contractId)))),
+          )
+        } else {
+          ctx.testingHelpers.failingExercise(
+            testIdsGenerator.testCommandId("SDVl9"),
+            submitter = delegate,
+            template = templateIds.delegation,
+            contractId = delegationEv.contractId,
+            choice = "LookupByKeyDelegated",
+            arg = Value(Value.Sum.Record(lookupArg(Some(delegatedEv.contractId)))),
+            Code.INVALID_ARGUMENT,
+            "Expected the submitter 'Charlie' to be in maintainers 'Bob'",
+          )
+        }
+      } yield (succeed)
+    }
+
+    "reject fetching an undisclosed contract" in allFixtures { ctx =>
```

Comment:
The notes from the previous test case apply to this as well.

Reply:
Addressed in 9bf50c0.